apps/user_ldap/lib/Service/UpdateGroupsService.php

```diff
@@ -0,0 +1,188 @@
+ * @author Côme Chilliet <come.chilliet@nextcloud.com>
+ * @author Joas Schilling <coding@schilljs.com>
+ * @author Lukas Reschke <lukas@statuscode.ch>
+ * @author Morris Jobke <hey@morrisjobke.de>
+ * @author Robin Appelman <robin@icewind.nl>
+ * @author Robin McCorkell <robin@mccorkell.me.uk>
+ *
+ * @license AGPL-3.0
+ *
+ * This code is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License, version 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License, version 3,
+ * along with this program. If not, see <http://www.gnu.org/licenses/>
+ *
+ */
+
+namespace OCA\User_LDAP\Service;
+
+use OCA\User_LDAP\Db\GroupMembership;
+use OCA\User_LDAP\Db\GroupMembershipMapper;
+use OCA\User_LDAP\Group_Proxy;
+use OCP\DB\Exception;
+use OCP\EventDispatcher\IEventDispatcher;
+use OCP\Group\Events\GroupCreatedEvent;
+use OCP\Group\Events\UserAddedEvent;
+use OCP\Group\Events\UserRemovedEvent;
+use OCP\IGroup;
+use OCP\IGroupManager;
+use OCP\IUser;
+use OCP\IUserManager;
+use Psr\Log\LoggerInterface;
+
+class UpdateGroupsService {
+	public function __construct(
+		private Group_Proxy $groupBackend,
+		private IEventDispatcher $dispatcher,
+		private IGroupManager $groupManager,
+		private IUserManager $userManager,
+		private LoggerInterface $logger,
+		private GroupMembershipMapper $groupMembershipMapper,
+	) {
+	}
+
+	/**
+	 * @throws Exception
+	 */
+	public function updateGroups(): void {
+		$knownGroups = $this->groupMembershipMapper->getKnownGroups();
+		$actualGroups = $this->groupBackend->getGroups();
+
+		if (empty($actualGroups) && empty($knownGroups)) {
+			$this->logger->info(
+				'service "updateGroups" – groups do not seem to be configured properly, aborting.',
+			);
+			return;
+		}
+
+		$this->handleKnownGroups(array_intersect($actualGroups, $knownGroups));
+		$this->handleCreatedGroups(array_diff($actualGroups, $knownGroups));
+		$this->handleRemovedGroups(array_diff($knownGroups, $actualGroups));
+
+		$this->logger->debug('service "updateGroups" – Finished.');
+	}
+
+	/**
+	 * @param string[] $groups
+	 * @throws Exception
+	 */
+	public function handleKnownGroups(array $groups): void {
+		$this->logger->debug('service "updateGroups" – Dealing with known Groups.');
+
+		foreach ($groups as $group) {
+			$this->logger->debug('service "updateGroups" – Dealing with {group}.', ['group' => $group]);
+			$groupMemberships = $this->groupMembershipMapper->findGroupMemberships($group);
+			$knownUsers = array_map(
+				fn (GroupMembership $groupMembership): string => $groupMembership->getUserid(),
```

Comment:
```suggestion
				static fn (GroupMembership $groupMembership): string => $groupMembership->getUserid(),
```
I think we can avoid binding `$this `